x/swaprouter/router_test.go

```diff
@@ -88,30 +245,77 @@ func (suite *KeeperTestSuite) TestMultihopSwapExactAmountIn() {
+				var records []poolincentivestypes.DistrRecord
+				totalWeight := sdk.NewInt(int64(len(tc.incentivizedGauges)))
+				for _, gauge := range tc.incentivizedGauges {
+					records = append(records, poolincentivestypes.DistrRecord{GaugeId: gauge, Weight: sdk.OneInt()})
+				}
+				distInfo := poolincentivestypes.DistrInfo{
+					TotalWeight: totalWeight,
+					Records:     records,
 				}
-				suite.App.PoolIncentivesKeeper.SetDistrInfo(suite.Ctx, test)
+				suite.App.PoolIncentivesKeeper.SetDistrInfo(suite.Ctx, distInfo)
 			}
 
-			tokenOut, err := swaprouterKeeper.RouteExactAmountIn(suite.Ctx, suite.TestAccs[0], tc.routes, tc.tokenIn, tc.tokenOutMinAmount)
+			// calcOutAmountAsSeparateSwaps calculates the multi-hop swap as separate swaps, while also
+			// utilizing a cacheContext so the state does not change
+			calcOutAmountAsSeparateSwaps := func(osmoFeeReduced bool) sdk.Coin {
+				cacheCtx, _ := suite.Ctx.CacheContext()
+
+				if osmoFeeReduced {
+					// extract route from swap
+					route := types.SwapAmountInRoutes(tc.routes)
+					// utilizing the extracted route, determine the routeSwapFee and sumOfSwapFees
+					// these two variables are used to calculate the overall swap fee utilizing the following formula
+					// swapFee = routeSwapFee * ((pool_fee) / (sumOfSwapFees))
+					routeSwapFee, sumOfSwapFees, err := swaprouterKeeper.GetOsmoRoutedMultihopTotalSwapFee(suite.Ctx, route)
+					suite.Require().NoError(err)
+					nextTokenIn := tc.tokenIn
+					for _, hop := range tc.routes {
+						// extract the current pool's swap fee
+						hopPool, err := gammKeeper.GetPoolAndPoke(cacheCtx, hop.PoolId)
+						suite.Require().NoError(err)
+						currentPoolSwapFee := hopPool.GetSwapFee(cacheCtx)
+						// utilize the routeSwapFee, sumOfSwapFees, and current pool swap fee to calculate the new reduced swap fee
+						swapFee := routeSwapFee.Mul((currentPoolSwapFee.Quo(sumOfSwapFees)))
+						// we then do individual swaps until we reach the end of the swap route
+						tokenOut, err := gammKeeper.SwapExactAmountIn(cacheCtx, suite.TestAccs[0], hopPool, nextTokenIn, hop.TokenOutDenom, sdk.OneInt(), swapFee)
+						suite.Require().NoError(err)
+						nextTokenIn = sdk.NewCoin(hop.TokenOutDenom, tokenOut)
+					}
+					return nextTokenIn
+				} else {
+					nextTokenIn := tc.tokenIn
+					for _, hop := range tc.routes {
+						hopPool, err := gammKeeper.GetPoolAndPoke(cacheCtx, hop.PoolId)
+						suite.Require().NoError(err)
+						updatedPoolSwapFee := hopPool.GetSwapFee(cacheCtx)
+						tokenOut, err := gammKeeper.SwapExactAmountIn(cacheCtx, suite.TestAccs[0], hopPool, nextTokenIn, hop.TokenOutDenom, sdk.OneInt(), updatedPoolSwapFee)
+						suite.Require().NoError(err)
+						nextTokenIn = sdk.NewCoin(hop.TokenOutDenom, tokenOut)
+					}
+					return nextTokenIn
+				}
+			}
 
 			if tc.expectError {
+				_, err := swaprouterKeeper.RouteExactAmountIn(suite.Ctx, suite.TestAccs[0], tc.routes, tc.tokenIn, tc.tokenOutMinAmount)
```

Comment:
We can move this up and share between if / else branches

Reply:
We can only do this if we use a cache context for routeExactAmount in as well, otherwise the second swap that happens happens with an unbalanced pool. I personally would rather use cache ctx for the expected and then run the actual swap after and compare the results. lmk what you think

Reply:
Ahh I see. Wasn't aware of this constraint. I'm fine with whatever you think is best